CodingStandard.md

```diff
@@ -0,0 +1,1425 @@
+between public and private members and methods, and I definitely dislike the 
+trailing `_` they use for class members.  What do you guys like?]
+
+**Header Guards**
+
+All headers should have `#define` guards that match the file name.  The `#endif`
+should be followed by a comment indicating the symbol.  For example,
+some_file.h would be guarded as follows:
+
+```c++
+#ifndef SOME_FILE_H
+#define SOME_FILE_H
+
+   ...
+
+#endif // SOME_FILE_H
+```
+
+[The #pragma once Macro accomplishes this same goal, but it is <a 
+
+href="https://en.wikipedia.org/wiki/Pragma_once">none-standard</a>.]
+
+**Header File Layout**
+
+A typical .h file should look like this:
+
+```c++
+/*
+ *  (Copyright and File Header)
+ */
+#ifndef FOOS_FOO_H
+#define FOOS_FOO_H
+
+namespace foos {
+
+/*
+ *  (Class Header)
+ */
+struct Widget
+{
+    int x{};
+    int y{};
+};
+
+/*
+ *  (Class Header)
+ */
+class Foo
+{
+public:
+    Foo();
+    explicit Foo(int bar);
+    Foo(int bar, int id);
+    Foo(Foo& other) = delete;
+    Foo(Foo&& other) = delete;
+    Foo& operator=(Foo& other) = delete;
+    Foo& operator=(Foo other) = delete;
+    ~Foo();
+    int bar() const
+    {
+        return bar_;
+    }
+    void set_bar(int bar)
+    {
+        bar_ = bar;
+    }
+    int double_bar() const;
+
+private:
+    void compute_bar(int x);
+    int bar_;
+    int id_;
+    Widget widget_;
+}
+
+/*
+ *  (Routine Header)
+ */
+inline int Foo::double_bar() const
+{
+    return 2 * bar_;
+}
+
+} // namespace foos
+
+#endif // FOOS_FOO_H
+```
+
+Sections should be ordered public, protected, then private.
+
+Class members should be declared in this order:
+ - Using-declarations, typedefs and enums
+ - Constants (static const/constexpr data members)
+ - Constructors and assignment operators
+ - Destructor
+ - Methods, including static methods
+ - Member variables (except constant data members)
+
+Function parameters should be declared in this order:
+  - Input
+  - Input/Output
+  - Output
+
+
+**Implementation File Layout**
+
+A typical .cpp file should look like this:
+
+```c++
+/*
+ *  (Copyright and File Header)
+ */
+#include "foo.h"
+#include "library.h"
+#include "other.h"
+#include "core.h"
+
+namespace foos {
+
+/*
+ *  (Routine Header)
+ */
+Foo::Foo() : bar_{}, foo_{}, widget_{}
+{
+}
+
+/*
+ *  (Routine Header)
+ */
+Foo::Foo(int bar): bar_{bar}, id_{}, widget_{}
+{
+}
+
+/*
+ *  (Routine Header)
+ */
+Foo::Foo(int bar, int id): bar_{bar}, id_{id}, widget_{}
+{
+}
+
+/*
+ *  (Routine Header)
+ */
+Foo::~Foo() = default;
+
+/*
+ *  (Routine Header)
+ */
+void Foo::compute_bar(int x)
+{
+    bar_ = 5 * x;
+}
+
+} // namespace foos
+
+```
+
+Member function definitions can appear in any order in the .cpp file, but
+should be ordered logically to help the reader.
+
+The #include for the class header should be the first substantive line in the
+source file, following the file header and the copyright block.
+
+Source code should be wrapped at 90 characters.
```

Comment:
I have seen 80 for legacy, but why 90?
Also tabs or spaces?

Reply:
I've always used 90.  No particular reason.

We had a server script that replaced tabs with spaces so that developers could use whichever they preferred when working.